pkg/controller/petset/pet_test.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package petset
+
+import (
+	"fmt"
+	"net/http/httptest"
+	"testing"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/testapi"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/fake"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/client/testing/core"
+	"k8s.io/kubernetes/pkg/runtime"
+	utiltesting "k8s.io/kubernetes/pkg/util/testing"
+)
+
+func makeTwoDifferntPCB() (pcb1, pcb2 *pcb) {
+	userAdded := api.Volume{
+		Name: "test",
+		VolumeSource: api.VolumeSource{
+			EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageMediumMemory},
+		},
+	}
+	ps := newStatefulSet(2)
+	pcb1, _ = newPCB("1", ps)
+	pcb2, _ = newPCB("2", ps)
+	pcb2.pod.Spec.Volumes = append(pcb2.pod.Spec.Volumes, userAdded)
+	return pcb1, pcb2
+}
+
+func TestUpdatePetWithoutRetry(t *testing.T) {
+	body := runtime.EncodeOrDie(testapi.Default.Codec(), &api.Pod{ObjectMeta: api.ObjectMeta{Name: "empty_pod"}})
+	fakeHandler := utiltesting.FakeHandler{
+		StatusCode:   200,
+		ResponseBody: string(body),
+	}
+	testServer := httptest.NewServer(&fakeHandler)
+	defer testServer.Close()
+
+	client := clientset.NewForConfigOrDie(&restclient.Config{Host: testServer.URL, ContentConfig: restclient.ContentConfig{GroupVersion: &registered.GroupOrDie(api.GroupName).GroupVersion}})
+	petClient := apiServerPetClient{
+		c: client,
+	}
+
+	pcb1, pcb2 := makeTwoDifferntPCB()
+	// invalid pet with empty pod
+	pcb3 := *pcb1
+	pcb3.pod = nil
+
+	testCases := []struct {
+		realPet     *pcb
+		expectedPet *pcb
+		errOK       func(error) bool
```

Comment:
hmm...

If use `expectErr bool`, then

I should check the result in the following style:

```
err := petClient.Update(test.realPet, test.expectedPet)
if err != nil && !expectErr || err ==nil && expectErr {
}
}
```